### PR TITLE
Webpack Redux

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

- Bundle entire application using Webpack (behind a MVT switch for now 😅 )
- All Webpack chunks are now hashed, and added to the asset map 🔀 
- Upgrade to Webpack 2.2.0 🎉 

## What is the value of this and can you measure success?

We can now conduct a more complete test of the Webpack-bundled application, without having to run the r.js-bundled app alongside it

## Does this affect other platforms - Amp, Apps, etc?

No

## Tested in CODE?

- [x] tested on CODE

**NB:** this PR has nothing to do with [Redux](http://redux.js.org/)

### CODE REVIEW

File: tools/webpack-loaders/svg-loader/index.js
Comment (SiAdcock) [inline tools/webpack-loaders/svg-loader/index.js]:
```
path.basename(this.resourcePath, '.svg');
```

https://nodejs.org/api/path.html#path_path_basename_path_ext

File: tools/webpack-loaders/svg-loader/index.js
Comment (sndrs) [inline tools/webpack-loaders/svg-loader/index.js]: 👍

File: static/src/javascripts/boot-webpack.js
Comment (NataliaLKB) [inline static/src/javascripts/boot-webpack.js]: Asked Simon in person about this. Would it be worth adding a comment explaining why this variable is in this format?

File: tools/__tasks__/compile/hash/index.js
Comment (NataliaLKB) [inline tools/__tasks__/compile/hash/index.js]: What do you guys think about this rule: https://github.com/airbnb/javascript#whitespace--after-blocks

Personally I like it. I find code much easier to read with nice whitespace. I also read [maintainable JS](http://dl.finebook.ir/book/4a/13598.pdf) which I found really useful.

**Comment by gtrufitt** [comment]
https://github.com/guardian/frontend/blob/03df2b81de80cc7ff4e0b50e4f5d048e0222fa00/common/app/views/fragments/analytics/google.scala.html#L39

You could use `window.config.tests && window.config.tests.abWebpackBundle ? 'webpack-load' : 'curl-load'` or do it on the serverside with the MVT switch.

**Review by gtrufitt — APPROVED** [review]
I'm not 100% on top of the architecture here, so it'd be good to have a quick explainer at some point, but lgtm. Awesome work! 💯

**Review by NataliaLKB — APPROVED** [review]
Amazing work guys! LGTM 👍

File: static/src/javascripts/boot-webpack.js
Comment (sndrs) [inline static/src/javascripts/boot-webpack.js]: if that wasn't clear, then absolutely - on its way

File: tools/__tasks__/compile/hash/index.js
Comment (sndrs) [inline tools/__tasks__/compile/hash/index.js]: weird that's not picked up in their eslint config (which this file should be subject to)?

if you want update the lint config to catch it (esp if it can fix them automatically) be totally happy to +1 a PR

but we should be cautious about picking up style prefs in PRs – there is just no way we're all going to agree (the repo's history is littered with 'discussions' around this 😉). v up for broad, super-strict linting, but one value of being super-strict is that whatever gets through should count as ok...

that said, it seems like it should be in the config (unless its' tricky to lint?)

Comment (sndrs) [inline tools/__tasks__/compile/hash/index.js]: maybe this? https://github.com/mgeraci/eslint-plugin-block-function-spacing

**Comment by sndrs** [comment]
will merge #15648 into this before merging, if possible

File: tools/webpack-loaders/svg-loader/index.js
Comment (gtrufitt) [inline tools/webpack-loaders/svg-loader/index.js]: I was reading this morning that this is the default in webpack 2, so not needed here.

Comment (SiAdcock) [inline tools/webpack-loaders/svg-loader/index.js]: Can you remember where you read this? Docs say loaders are [not cacheable by default](https://webpack.js.org/api/loaders/#cacheable).

Comment (SiAdcock) [inline tools/webpack-loaders/svg-loader/index.js]: Plus ça change...

Comment (gtrufitt) [inline tools/webpack-loaders/svg-loader/index.js]: cool, cool...

image

**Pull request merged**
